shell/platform/android/android_image_generator.cc

```diff
@@ -0,0 +1,192 @@
+
+#include "flutter/shell/platform/android/android_image_generator.h"
+
+#include <android/bitmap.h>
+#include <android/hardware_buffer.h>
+
+#include "flutter/fml/platform/android/jni_util.h"
+
+namespace flutter {
+
+static fml::jni::ScopedJavaGlobalRef<jclass>* g_flutter_jni_class = nullptr;
+static jmethodID g_decode_image_method = nullptr;
+
+AndroidImageGenerator::~AndroidImageGenerator() = default;
+
+AndroidImageGenerator::AndroidImageGenerator(sk_sp<SkData> data)
+    : data_(data), image_info_(SkImageInfo::MakeUnknown()) {}
+
+const SkImageInfo& AndroidImageGenerator::GetInfo() {
+  header_decoded_latch_.Wait();
+  return image_info_;
+}
+
+unsigned int AndroidImageGenerator::GetFrameCount() const {
+  return 1;
+}
+
+unsigned int AndroidImageGenerator::GetPlayCount() const {
+  return 1;
+}
+
+const ImageGenerator::FrameInfo AndroidImageGenerator::GetFrameInfo(
+    unsigned int frame_index) const {
+  return {.required_frame = std::nullopt,
+          .duration = 0,
+          .disposal_method = SkCodecAnimation::DisposalMethod::kKeep};
+}
+
+SkISize AndroidImageGenerator::GetScaledDimensions(float desired_scale) {
+  return GetInfo().dimensions();
+}
+
+bool AndroidImageGenerator::GetPixels(const SkImageInfo& info,
+                                      void* pixels,
+                                      size_t row_bytes,
+                                      unsigned int frame_index,
+                                      std::optional<unsigned int> prior_frame) {
+  fully_decoded_latch_.Wait();
+
+  if (kRGBA_8888_SkColorType != info.colorType()) {
+    return false;
+  }
+
+  switch (info.alphaType()) {
+    case kOpaque_SkAlphaType:
+      if (kOpaque_SkAlphaType != GetInfo().alphaType()) {
+        return false;
+      }
+      break;
+    case kPremul_SkAlphaType:
+      break;
+    default:
+      return false;
+  }
+
+  // Since this decoder is only installed for API 28 and 29, there is no
+  // supported way to work around this unfortunate copy. It's possible to
+  // produce an `SkImage` direcly from an `AHardwareBuffer`, but
+  // `AndroidBitmap_getHardwareBuffer` and `Bitmap.getHardwareBuffer` are only
+  // available in API 30+ and 31+ respectively.
+  memcpy(pixels, software_decoded_data_->data(),
+         software_decoded_data_->size());
+  return true;
+}
+
+void AndroidImageGenerator::DecodeImage() {
+  DoDecodeImage();
+
+  header_decoded_latch_.Signal();
+  fully_decoded_latch_.Signal();
```

Comment:
For threading here: won't this deadlock if someone call GetInfo (and other methods that call GetInfo) before calling DecodeImage?

Reply:
Ah, so `AndroidImageGenerator::MakeFromData` factory is the only reasonable way to use this; it kicks off the decoding immediately in a separate task. And so I'm just gonna make the constructor private.